File: vehicle-tracker-backend/vehicle-service/src/main/java/com/github/diogocerqueiralima/application/mappers/VehicleMapper.java
package com.github.diogocerqueiralima.application.mappers;

import com.github.diogocerqueiralima.application.results.VehicleResult;
import com.github.diogocerqueiralima.domain.model.Vehicle;
import org.springframework.stereotype.Component;

@Component("vm-application")
public class VehicleMapper {

    public VehicleResult toResult(Vehicle vehicle) {
        return new VehicleResult(
                vehicle.getId(),
                vehicle.getVin(),
                vehicle.getPlate(),
                vehicle.getModel(),
                vehicle.getManufacturer(),
                vehicle.getYear(),
                vehicle.getOwnerId()
        );
    }

    public Vehicle toDomain(VehicleResult vehicleResult) {
        return new Vehicle(
                vehicleResult.id(),
                vehicleResult.vin(),
                vehicleResult.plate(),
                vehicleResult.model(),
                vehicleResult.manufacturer(),
                vehicleResult.year(),
                vehicleResult.ownerId()
        );
    }

}

File: vehicle-tracker-mobile/app/src/main/java/com/github/diogocerqueiralima/infrastructure/vehicles/mappers/VehicleMapper.kt
package com.github.diogocerqueiralima.infrastructure.vehicles.mappers

import com.github.diogocerqueiralima.domain.vehicles.model.Vehicle
import com.github.diogocerqueiralima.infrastructure.vehicles.entities.VehicleEntity
import java.time.LocalDate

/**
 * Extension function to convert a [VehicleEntity] infrastructure entity to a [Vehicle] domain model.
 */
fun VehicleEntity.toDomain() = Vehicle(
    vin = this.vin,
    plate = this.plate,
    model = this.model,
    manufacturer = this.manufacturer,
    manufacturingDate = LocalDate.parse(this.manufacturingDate)
)